### PR TITLE
Feature: PM-511 parametric gaits in rqt

### PR DESCRIPTION
**Pull request by RutgerVanBeek**
Closes PM-511

## Description

I have made added an extra option to the subgait versiosn to choose from, called "parametric". When choosing this version a pop-up will appear. In this pop-up one can choose a parameter and versions for the parametric gait. This is then stored as a versions string with the following format: '_{parameter}_({base version})_({other version})'. 
## Changes

* Added a pop up menu for creating a parametric gait. 

## Testing
Test this with the corresponding branch in the March repo, otherwise it will not be able to find files with this version.

### CODE REVIEW

File: march_rqt_gait_selection/src/march_rqt_gait_selection/gait_selection_view.py [hunk 1]
```diff
@@ -157,6 +165,17 @@ def update_version_colors(self):
...
+                        versions = self.available_gaits[gait_name][subgait_name]
+                        if self._parametric_pop_up.show_pop_up(versions):
+                            new_version = self.get_parametric_version()
+                            subgait_label.setStyleSheet('color:{color}'.format(color=self._colors['warning']))
+                            subgait_menu.addItem(new_version)
+                            subgait_menu.setCurrentIndex(subgait_menu.count() - 1)
+                        else:
+                            # parametric pop up window unsuccessful stopped, reset version to default
+                            current_version_index = versions.index(self.version_map[gait_name][subgait_name])
+                            subgait_menu.setCurrentIndex(current_version_index - 1)
```
Comment (Olavhaasie) [inline march_rqt_gait_selection/src/march_rqt_gait_selection/gait_selection_view.py]: When the `current_version_index` is 0 you get an out of bounds. You can try by choosing parametric on a subgait with only 1 subgait version and then closing it.

File: march_rqt_gait_selection/src/march_rqt_gait_selection/parametric_pop_up.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
+
+from PyQt5.QtCore import Qt
+from PyQt5.QtWidgets import QComboBox, QDialog, QDialogButtonBox, QFormLayout, QGridLayout, QLabel, QLineEdit,\
+    QMessageBox
+
+
+class ParametricPopUpWindow(QDialog):
```
Comment (Olavhaasie) [inline march_rqt_gait_selection/src/march_rqt_gait_selection/parametric_pop_up.py:7]: You could also create a `.ui` file for this, just like the main widget and load it.

**Review by Olavhaasie — CHANGES_REQUESTED** [review]
Looks good! I think in the future, when this could be expanded another way of handling the parameters will be necessary. For example, if you want more parameters the version string would become very long.

Also another small point. Perhaps you could add a small description next to the parameter input field, to describe what it does?

File: march_rqt_gait_selection/src/march_rqt_gait_selection/gait_selection_view.py [hunk 3]
```diff
@@ -263,3 +282,11 @@ def _show_version_map_pop_up(self):
             version_map_string += '\n'
 
         self._version_map_pop_up.show_message(version_map_string)
+
+    def _show_parametric_pop_up(self):
+        """Use a pop up window to get the base version, other version and parameter for a parametric subgait."""
+        self._parametric_pop_up.show_pop_up()
```
Comment (Joris-JW) [inline march_rqt_gait_selection/src/march_rqt_gait_selection/gait_selection_view.py]: should this function not have the `versions` variable populated?

**Review by Joris-JW — CHANGES_REQUESTED** [review]
since one of the drop-down options triggers a pop up menu maybe it is nice to disable the scrolling through the drop-down with the mouse wheel, just to prevent sudden popups. furthermore, looks good and works nice :fire: 

And what olav said might be nice (about thedescription) and i think the quickest way would be to add a tooltip when dragging the mouse over the parameter label. This is very easy to implement using a `.ui` file

File: march_rqt_gait_selection/resource/parametric_pop_up.ui [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+      </property>
+      <property name="text">
+       <string>base version</string>
+      </property>
+     </widget>
+    </item>
+    <item row="1" column="1">
+     <widget class="QComboBox" name="baseVersionComboBox">
+      <property name="toolTip">
+       <string extracomment="The parametric subgait would result to this version if the parameter is 0%.">The parametric subgait would result to this version if the parameter is 0%.</string>
```
Comment (Olavhaasie) [inline march_rqt_gait_selection/resource/parametric_pop_up.ui]:
```suggestion
       <string extracomment="The parametric subgait would result in this version if the parameter is 0%.">The parametric subgait would result in this version if the parameter is 0%.</string>
```

File: march_rqt_gait_selection/resource/parametric_pop_up.ui [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+      </property>
+     </widget>
+    </item>
+    <item row="2" column="0">
+     <widget class="QLabel" name="otherVersionLabel">
+      <property name="toolTip">
+       <string extracomment="The parametric subgait would result to this version if the parameter is 100%.">The parametric subgait would result to this version if the parameter is 100%.</string>
```
Comment (Olavhaasie) [inline march_rqt_gait_selection/resource/parametric_pop_up.ui]:
```suggestion
       <string extracomment="The parametric subgait would result in this version if the parameter is 100%.">The parametric subgait would result in this version if the parameter is 100%.</string>
```

**Review by Olavhaasie — CHANGES_REQUESTED** [review]
I don't know why but my popup looks like this. It could have something to do with my resolution, but the normal rqt plugin is fine.

image

Also a box and slider would be nice for inputting the parameter. Similar to what we use for the time slider in the gait generation. Also, you wouldn't have to update the label every time.

**Review by Olavhaasie — APPROVED** [review]
Everything looks good now :+1:

If you want to and if have some time left you could take a look at my other suggestion :innocent: 

> Also a box and slider would be nice for inputting the parameter. Similar to what we use for the time slider in the gait generation. Also, you wouldn't have to update the label every time.

**Comment by RutgerVanBeek** [comment]
> Also a box and slider would be nice for inputting the parameter. Similar to what we use for the time slider in the gait generation. Also, you wouldn't have to update the label every time.

What is the difference with the current slider. What would you like to see different?
I think updating the label is nice, to see the exact value you chose. (Say you want to do exactly 50%).

**Comment by Olavhaasie** [comment]
> > Also a box and slider would be nice for inputting the parameter. Similar to what we use for the time slider in the gait generation. Also, you wouldn't have to update the label every time.
> 
> What is the difference with the current slider. What would you like to see different?
> I think updating the label is nice, to see the exact value you chose. (Say you want to do exactly 50%).

I had this idea of using a input box and a slider, so that you could enter 0.5 exactly without having to slide it, or just estimate a value with the slider. But it's not that important, so nevermind unless you want to implement it yourself.

**Review by Joris-JW — APPROVED** [review]
:100:

**Pull request merged**